libirods_rule_engine_plugin-elasticsearch.cpp

```diff
@@ -1,913 +1,841 @@
+		                           .arg5 = const_cast<char*>(_units.c_str())};
+
+		auto status = rsModAVUMetadata(_rei->rsComm, &set_op);
+		if (status < 0) {
+			THROW(status, boost::format("failed to update object [%s] metadata") % _object_path);
+		}
+	} // update_object_metadata
+
+	void invoke_indexing_event_full_text(ruleExecInfo_t* _rei,
+	                                     const std::string& _object_path,
+	                                     const std::string& _source_resource,
+	                                     const std::string& _index_name)
+	{
+		try {
+			if (config->bulk_count_ < 0) {
+				// TODO Do we protect against this?
+			}
+
+			std::string doc_type{"text"}; // TODO What is this for?
+			apply_document_type_policy(_rei, _object_path, _source_resource, &doc_type);
+
+			const std::string object_id = get_object_index_id(_rei, _object_path);
+			std::vector<char> buffer(config->read_size_);
+			irods::experimental::io::server::basic_transport<char> xport(*_rei->rsComm);
+			irods::experimental::io::idstream in{xport, _object_path};
+
+			int chunk_counter{0};
+			bool need_final_perform{false};
+			std::stringstream ss;
+
+			while (in) {
+				in.read(buffer.data(), buffer.size());
+
+				// The indexing instruction.
+				// clang-format off
+				ss << json{{"index", {
+					{"_id", fmt::format("{}_{}", object_id, chunk_counter++)}
+				}}}.dump() << '\n';
+				// clang-format on
+
+				// The defaults for the .dump() member function.
+				constexpr int indent = -1;
+				constexpr char indent_char = ' ';
+				constexpr bool ensure_ascii = false;
+
+				// The data to index.
+				// The version of .dump() invoked here instructs the library to ignore
+				// invalid UTF-8 sequences. All bytes are copied to the output unchanged.
+				// clang-format off
+				ss << json{
+					{"absolutePath", _object_path},
+					{"data", std::string_view(buffer.data(), in.gcount())}
+				}.dump(indent, indent_char, ensure_ascii, json::error_handler_t::ignore) << '\n';
+				// clang-format on
+
+				// Send bulk request if chunk counter has reached bulk limit.
+				if (chunk_counter == config->bulk_count_) {
+					chunk_counter = 0;
+
+					const auto res =
+						send_http_request(config->hosts_[0], http::verb::post, _index_name + "/_bulk", ss.str());
+
+					if (!res.has_value()) {
+						rodsLog(LOG_ERROR, "%s: No response from elasticsearch host.", __func__);
+					}
+					else {
+						rodsLog(LOG_ERROR,
+						        "%s: Error sending request to elasticsearch host. [http_status_code=[%d]]",
+						        __func__,
+						        res->result_int());
+					}
+
+					ss.str("");
+				}
+			}
+
+			if (chunk_counter > 0) {
+				// Elasticsearch limits the maximum size of a HTTP request to 100mb.
+				// See https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html.
+				const auto res =
+					send_http_request(config->hosts_[0], http::verb::post, _index_name + "/_bulk", ss.str());
```

Comment:
Remove this.

Elasticsearch should return information about these kinds of failures.

I thought about logging a message if the http request body exceeds 100mb, but decided against doing that because the ES server may support increasing the allowed size of an http request.

Reply:
I've decided to leave this in.

See https://github.com/irods/irods_capability_indexing/pull/131#discussion_r1493777122.